File: Sources/HerdrMobile/Console/AttachLinkIndex.swift
import Foundation
import Observation

/// One exact HTTP(S) target observed during the current Attach.
struct AttachLink: Identifiable, Equatable {
    let target: String
    let url: URL

    var id: String { target }
    var host: String { url.host ?? "" }
}

/// Owns the bounded, memory-only Attach Link collection and incremental
/// discovery state. A terminal pipeline may be replaced without replacing
/// this index; the enclosing `AgentAttachStore` clears it on leave.
@MainActor
@Observable
final class AttachLinkIndex {
    private static let maximumLinkCount = 20

    private(set) var links: [AttachLink] = []

    @ObservationIgnored private let onDistinctLink: (AttachLink) -> Void
    private var scanner = AttachLinkScanner()

    init(onDistinctLink: @escaping (AttachLink) -> Void) {
        self.onDistinctLink = onDistinctLink
    }

    func receive(_ data: Data) {
        var scanner = scanner
        scanner.receive(data) { target in
            record(target)
        }
        self.scanner = scanner
    }

    /// Supplements stream discovery from Ghostty's current visible text.
    /// Every snapshot is independent: its newlines remain hard boundaries
    /// because the embedding API does not identify visual soft wraps.
    func receiveViewportText(_ text: String) {
        var viewportScanner = AttachLinkScanner()
        viewportScanner.receive(Data(text.utf8)) { target in
            record(target, moveExistingToFront: false)
        }
        viewportScanner.finishOutput { target in
            record(target, moveExistingToFront: false)
        }
    }

    func finishOutput() {
        var scanner = scanner
        scanner.finishOutput { target in
            record(target)
        }
        self.scanner = scanner
    }

    func clear() {
        scanner = AttachLinkScanner()
        links.removeAll(keepingCapacity: false)
    }

    private func record(_ target: String, moveExistingToFront: Bool = true) {
        guard let url = TerminalLinkPolicy.url(for: target) else { return }
        let isDistinct = !links.contains(where: { $0.target == target })
        if !moveExistingToFront, !isDistinct {
            return
        }
        let link = AttachLink(target: target, url: url)
        links.removeAll { $0.target == link.target }
        links.insert(link, at: 0)
        if links.count > Self.maximumLinkCount {
            links.removeLast(links.count - Self.maximumLinkCount)
        }
        if isDistinct {
            onDistinctLink(link)
        }
    }
}

/// Removes terminal control content while preserving stream state across
/// arbitrary PTY chunks. SGR styling is transparent to visible URLs; other
/// controls end plain candidates. OSC 8 targets are collected explicitly so
/// their labels cannot hide or replace the real destination.
private struct AttachLinkScanner {
    private enum State: Equatable {
        case text
        case escape
        case escapeSequence
        case csi
        case osc
        case oscEscape
        case stringControl
        case stringControlEscape
    }

    private var state = State.text
    private var plain = PlainAttachLinkScanner()
    private var osc = OSC8Accumulator()
    private var isOSC8LabelActive = false

    mutating func receive(_ data: Data, record: (String) -> Void) {
        for byte in data {
            receive(byte, record: record)
        }
    }

    mutating func finishOutput(record: (String) -> Void) {
        if state != .text {
            finishPlainCandidate(record: record)
        } else if !isOSC8LabelActive, let target = plain.finishOutput() {
            record(target)
        }
        state = .text
        osc.reset()
        isOSC8LabelActive = false
    }

    private mutating func receive(_ byte: UInt8, record: (String) -> Void) {
        switch state {
        case .text:
            receiveText(byte, record: record)

        case .escape:
            receiveAfterEscape(byte, record: record)

        case .escapeSequence:
            if (0x30...0x7E).contains(byte) {
                state = .text
            } else if byte == 0x1B {
                state = .escape
            }

        case .csi:
            if (0x40...0x7E).contains(byte) {
                if byte != 0x6D {
                    finishPlainCandidate(record: record)
                }
                state = .text
            } else if byte == 0x1B {
                finishPlainCandidate(record: record)
                state = .escape
            }

        case .osc:
            if byte == 0x07 || byte == 0x9C {
                finishOSC(record: record)
            } else if byte == 0x1B {
                state = .oscEscape
            } else {
                osc.receive(byte)
            }

        case .oscEscape:
            if byte == 0x5C {
                finishOSC(record: record)
            } else {
                osc.invalidate()
                state = byte == 0x1B ? .oscEscape : .osc
            }

        case .stringControl:
            if byte == 0x9C {
                state = .text
            } else if byte == 0x1B {
                state = .stringControlEscape
            }

        case .stringControlEscape:
            state = byte == 0x5C ? .text : .stringControl
        }
    }

    private mutating func receiveText(_ byte: UInt8, record: (String) -> Void) {
        switch byte {
        case 0x1B:
            state = .escape
        case 0x9B:
            state = .csi
        case 0x9D:
            osc.reset()
            state = .osc
        case 0x90, 0x98, 0x9E, 0x9F:
            finishPlainCandidate(record: record)
            state = .stringControl
        case 0x80...0x9F:
            finishPlainCandidate(record: record)
        default:
            guard !isOSC8LabelActive else { return }
            plain.receive(byte, record: record)
        }
    }

    private mutating func receiveAfterEscape(
        _ byte: UInt8,
        record: (String) -> Void
    ) {
        switch byte {
        case 0x5B:
            state = .csi
        case 0x5D:
            osc.reset()
            state = .osc
        case 0x50, 0x58, 0x5E, 0x5F:
            finishPlainCandidate(record: record)
            state = .stringControl
        case 0x20...0x2F:
            finishPlainCandidate(record: record)
            state = .escapeSequence
        default:
            finishPlainCandidate(record: record)
            state = .text
        }
    }

    private mutating func finishOSC(record: (String) -> Void) {
        defer {
            osc.reset()
            state = .text
        }
        guard let hyperlink = osc.hyperlink else {
            finishPlainCandidate(record: record)
            return
        }

        finishPlainCandidate(record: record)
        switch hyperlink {
        case .close:
            isOSC8LabelActive = false
        case .open(let target):
            isOSC8LabelActive = true
            if let target {
                record(target)
            }
        }
    }

    private mutating func finishPlainCandidate(record: (String) -> Void) {
        guard !isOSC8LabelActive, let target = plain.finishOutput() else { return }
        record(target)
    }
}

private struct OSC8Accumulator {
    enum Hyperlink {
        case open(String?)
        case close
    }

    private enum Field: Equatable {
        case command
        case parameters
        case target
        case ignored
    }

    private static let maximumTargetBytes = 32 * 1024
    private static let maximumCommandBytes = 16

    private var field = Field.command
    private var command = Data()
    private var target = Data()
    private var isTargetOversized = false

    var hyperlink: Hyperlink? {
        guard field == .target else { return nil }
        guard !target.isEmpty else { return .close }
        guard !isTargetOversized,
            let target = String(data: target, encoding: .utf8)
        else {
            return .open(nil)
        }
        return .open(target)
    }

    mutating func receive(_ byte: UInt8) {
        switch field {
        case .command:
            if byte == 0x3B {
                field = command == Data("8".utf8) ? .parameters : .ignored
            } else if command.count < Self.maximumCommandBytes {
                command.append(byte)
            } else {
                field = .ignored
            }

        case .parameters:
            if byte == 0x3B {
                field = .target
            }

        case .target:
            guard !isTargetOversized else { return }
            guard target.count < Self.maximumTargetBytes else {
                target.removeAll(keepingCapacity: false)
                isTargetOversized = true
                return
            }
            target.append(byte)

        case .ignored:
            break
        }
    }

    mutating func invalidate() {
        field = .ignored
        command.removeAll(keepingCapacity: false)
        target.removeAll(keepingCapacity: false)
        isTargetOversized = false
    }

    mutating func reset() {
        field = .command
        command.removeAll(keepingCapacity: true)
        target.removeAll(keepingCapacity: true)
        isTargetOversized = false
    }
}

/// Recognizes plain web targets from printable terminal text. Network chunk
/// boundaries are invisible here; actual whitespace terminates candidates.
/// Both state buffers have hard upper bounds.
private struct PlainAttachLinkScanner {
    private static let maximumTargetBytes = 32 * 1024
    private static let longestSchemeBytes = Array("https://".utf8)
    private static let webSchemes = [
        Array("http://".utf8),
        Array("https://".utf8),
    ]
    private static let trailingSentencePunctuation = CharacterSet(charactersIn: ".,;:!")

    private var recent = Data()
    private var candidate = Data()
    private var isOversized = false

    mutating func receive(_ data: Data, record: (String) -> Void) {
        for byte in data {
            receive(byte, record: record)
        }
    }

    mutating func receive(_ byte: UInt8, record: (String) -> Void) {
        if Self.isBoundary(byte) {
            if let target = finishCandidate() {
                record(target)
            }
            recent.removeAll(keepingCapacity: true)
            return
        }

        if !candidate.isEmpty || isOversized {
            appendToCandidate(byte)
            return
        }

        recent.append(byte)
        if recent.count > Self.longestSchemeBytes.count {
            recent.removeFirst(recent.count - Self.longestSchemeBytes.count)
        }
        guard let scheme = Self.webSchemes.first(where: { recentEnds(with: $0) })
        else { return }
        candidate = Data(recent.suffix(scheme.count))
        recent.removeAll(keepingCapacity: true)
    }

    mutating func finishOutput() -> String? {
        defer { recent.removeAll(keepingCapacity: true) }
        return finishCandidate()
    }

    private mutating func appendToCandidate(_ byte: UInt8) {
        guard !isOversized else { return }
        guard candidate.count < Self.maximumTargetBytes else {
            candidate.removeAll(keepingCapacity: false)
            isOversized = true
            return
        }
        candidate.append(byte)
    }

    private mutating func finishCandidate() -> String? {
        defer {
            candidate.removeAll(keepingCapacity: true)
            isOversized = false
        }
        guard !isOversized, let text = String(data: candidate, encoding: .utf8) else {
            return nil
        }
        let target = Self.removingSurroundingPunctuation(from: text)
        return target.isEmpty ? nil : target
    }

    private func recentEnds(with suffix: [UInt8]) -> Bool {
        guard recent.count >= suffix.count else { return false }
        return zip(recent.suffix(suffix.count), suffix).allSatisfy {
            Self.lowercasedASCII($0.0) == $0.1
        }
    }

    private static func lowercasedASCII(_ byte: UInt8) -> UInt8 {
        (0x41...0x5A).contains(byte) ? byte + 0x20 : byte
    }

    private static func removingSurroundingPunctuation(from text: String) -> String {
        var target = text
        var removedCharacter = true
        while removedCharacter, let last = target.last {
            removedCharacter = false
            if last.unicodeScalars.allSatisfy(trailingSentencePunctuation.contains) {
                target.removeLast()
                removedCharacter = true
                continue
            }
            for (opening, closing) in [("(", ")"), ("[", "]"), ("{", "}")] {
                guard last == Character(closing) else { continue }
                let openingCount = target.reduce(into: 0) {
                    if $1 == Character(opening) { $0 += 1 }
                }
                let closingCount = target.reduce(into: 0) {
                    if $1 == Character(closing) { $0 += 1 }
                }
                if closingCount > openingCount {
                    target.removeLast()
                    removedCharacter = true
                }
                break
            }
        }
        return target
    }

    private static func isBoundary(_ byte: UInt8) -> Bool {
        byte <= 0x20 || byte == 0x7F
            || byte == 0x22 || byte == 0x27
            || byte == 0x3C || byte == 0x3E
    }
}

File: android/app/src/main/java/dev/bybee/heeler/detail/AttachLinkIndex.kt
package dev.bybee.heeler.detail

import java.net.URI

/** One ordinary HTTP(S) target observed during the current Attach session. */
data class AttachLink(val target: String) {
    val host: String = runCatching { URI(target).host.orEmpty() }.getOrDefault("")
}

/**
 * Bounded, memory-only index of links observed in an Attach PTY stream.
 *
 * The scanner keeps its state across arbitrary network chunks, strips terminal
 * control sequences, and explicitly records OSC 8 targets. The store owns this
 * index for one detail session and clears it when that detail is left.
 */
class AttachLinkIndex {
    private val scanner = AttachLinkScanner(::recordStreamTarget)
    private val streamTargets = mutableSetOf<String>()
    private val viewportOnlyTargets = mutableSetOf<String>()
    private val indexed = ArrayList<AttachLink>(MAXIMUM_LINK_COUNT)

    val links: List<AttachLink>
        get() = indexed.toList()

    fun receive(data: ByteArray) = scanner.receive(data)

    /**
     * Supplements stream discovery from a visible-terminal snapshot. Each
     * snapshot has hard line boundaries because soft wraps are unknowable here.
     */
    fun receiveViewportText(text: String) {
        val viewportScanner = AttachLinkScanner(::recordViewportTarget)
        viewportScanner.receive(text.encodeToByteArray())
        viewportScanner.finishOutput()
    }

    fun finishOutput() = scanner.finishOutput()

    fun clear() {
        scanner.reset()
        streamTargets.clear()
        viewportOnlyTargets.clear()
        indexed.clear()
    }

    private fun recordStreamTarget(target: String) {
        if (!isWebUrl(target)) return
        removeViewportOnlyPrefixes(target)
        if (!record(target, moveExistingToFront = true)) return
        streamTargets += target
        viewportOnlyTargets -= target
    }

    private fun recordViewportTarget(target: String) {
        if (!isWebUrl(target)) return
        if (indexed.any { it.target != target && it.target.startsWith(target) }) return
        removeViewportOnlyPrefixes(target)
        if (!record(target, moveExistingToFront = false)) return
        if (target !in streamTargets) viewportOnlyTargets += target
    }

    private fun removeViewportOnlyPrefixes(target: String) {
        val prefixes = viewportOnlyTargets.filter { it != target && target.startsWith(it) }.toSet()
        if (prefixes.isEmpty()) return
        indexed.removeAll { it.target in prefixes }
        viewportOnlyTargets.removeAll(prefixes)
    }

    private fun record(target: String, moveExistingToFront: Boolean): Boolean {
        if (!isWebUrl(target)) return false
        val existingIndex = indexed.indexOfFirst { it.target == target }
        if (!moveExistingToFront && existingIndex >= 0) return true
        if (existingIndex >= 0) indexed.removeAt(existingIndex)
        indexed.add(0, AttachLink(target))
        while (indexed.size > MAXIMUM_LINK_COUNT) indexed.removeAt(indexed.lastIndex)
        val retained = indexed.asSequence().map(AttachLink::target).toSet()
        streamTargets.retainAll(retained)
        viewportOnlyTargets.retainAll(retained)
        return true
    }

    private fun isWebUrl(target: String): Boolean =
        target.length <= MAXIMUM_TARGET_BYTES && runCatching {
            val uri = URI(target)
            val scheme = uri.scheme
            (scheme.equals("http", ignoreCase = true) || scheme.equals("https", ignoreCase = true)) &&
                !uri.host.isNullOrEmpty()
        }.getOrDefault(false)

    private class AttachLinkScanner(private val record: (String) -> Unit) {
        private enum class State { TEXT, ESCAPE, ESCAPE_SEQUENCE, CSI, OSC, OSC_ESCAPE, STRING_CONTROL, STRING_CONTROL_ESCAPE }

        private var state = State.TEXT
        private val plain = PlainAttachLinkScanner()
        private val osc8 = Osc8Accumulator()
        private var osc8LabelActive = false

        fun receive(data: ByteArray) {
            data.forEach(::receive)
        }

        fun finishOutput() {
            if (state != State.TEXT) {
                finishPlainCandidate()
            } else if (!osc8LabelActive) {
                plain.finishOutput()?.let(record)
            }
            reset()
        }

        fun reset() {
            state = State.TEXT
            plain.reset()
            osc8.reset()
            osc8LabelActive = false
        }

        private fun receive(byte: Byte) {
            val value = byte.toInt() and 0xFF
            when (state) {
                State.TEXT -> receiveText(value)
                State.ESCAPE -> receiveAfterEscape(value)
                State.ESCAPE_SEQUENCE -> when {
                    value in 0x30..0x7E -> state = State.TEXT
                    value == ESC -> state = State.ESCAPE
                }
                State.CSI -> when {
                    value in 0x40..0x7E -> {
                        if (value != 'm'.code) finishPlainCandidate()
                        state = State.TEXT
                    }
                    value == ESC -> {
                        finishPlainCandidate()
                        state = State.ESCAPE
                    }
                }
                State.OSC -> when (value) {
                    BEL, ST -> finishOsc()
                    ESC -> state = State.OSC_ESCAPE
                    else -> osc8.receive(value)
                }
                State.OSC_ESCAPE -> if (value == '\\'.code) {
                    finishOsc()
                } else {
                    osc8.invalidate()
                    state = if (value == ESC) State.OSC_ESCAPE else State.OSC
                }
                State.STRING_CONTROL -> when (value) {
                    ST -> state = State.TEXT
                    ESC -> state = State.STRING_CONTROL_ESCAPE
                }
                State.STRING_CONTROL_ESCAPE -> state = if (value == '\\'.code) State.TEXT else State.STRING_CONTROL
            }
        }

        private fun receiveText(value: Int) {
            when (value) {
                ESC -> state = State.ESCAPE
                CSI -> state = State.CSI
                OSC -> {
                    osc8.reset()
                    state = State.OSC
                }
                DCS, SOS, PM, APC -> {
                    finishPlainCandidate()
                    state = State.STRING_CONTROL
                }
                in 0x80..0x9F -> finishPlainCandidate()
                else -> if (!osc8LabelActive) plain.receive(value)?.let(record)
            }
        }

        private fun receiveAfterEscape(value: Int) {
            when (value) {
                '['.code -> state = State.CSI
                ']'.code -> {
                    osc8.reset()
                    state = State.OSC
                }
                'P'.code, 'X'.code, '^'.code, '_'.code -> {
                    finishPlainCandidate()
                    state = State.STRING_CONTROL
                }
                in 0x20..0x2F -> {
                    finishPlainCandidate()
                    state = State.ESCAPE_SEQUENCE
                }
                else -> {
                    finishPlainCandidate()
                    state = State.TEXT
                }
            }
        }

        private fun finishOsc() {
            val hyperlink = osc8.hyperlink()
            osc8.reset()
            state = State.TEXT
            if (hyperlink == null) {
                finishPlainCandidate()
                return
            }
            finishPlainCandidate()
            when (hyperlink) {
                is Osc8Accumulator.Hyperlink.Close -> osc8LabelActive = false
                is Osc8Accumulator.Hyperlink.Open -> {
                    osc8LabelActive = true
                    hyperlink.target?.let(record)
                }
            }
        }

        private fun finishPlainCandidate() {
            if (!osc8LabelActive) plain.finishOutput()?.let(record)
        }

        private companion object {
            const val ESC = 0x1B
            const val BEL = 0x07
            const val ST = 0x9C
            const val CSI = 0x9B
            const val OSC = 0x9D
            const val DCS = 0x90
            const val SOS = 0x98
            const val PM = 0x9E
            const val APC = 0x9F
        }
    }

    private class Osc8Accumulator {
        sealed interface Hyperlink {
            data class Open(val target: String?) : Hyperlink
            data object Close : Hyperlink
        }

        private enum class Field { COMMAND, PARAMETERS, TARGET, IGNORED }

        private var field = Field.COMMAND
        private val command = ArrayList<Byte>(MAXIMUM_COMMAND_BYTES)
        private var target = ByteArray(0)
        private var targetSize = 0
        private var targetOversized = false

        fun receive(value: Int) {
            when (field) {
                Field.COMMAND -> when (value) {
                    ';'.code -> field = if (command.toByteArray().decodeToString() == "8") Field.PARAMETERS else Field.IGNORED
                    else -> if (command.size < MAXIMUM_COMMAND_BYTES) command += value.toByte() else field = Field.IGNORED
                }
                Field.PARAMETERS -> if (value == ';'.code) field = Field.TARGET
                Field.TARGET -> appendTarget(value.toByte())
                Field.IGNORED -> Unit
            }
        }

        fun hyperlink(): Hyperlink? {
            if (field != Field.TARGET) return null
            if (targetSize == 0) return Hyperlink.Close
            if (targetOversized) return Hyperlink.Open(null)
            return runCatching { Hyperlink.Open(target.copyOf(targetSize).decodeToString()) }
                .getOrElse { Hyperlink.Open(null) }
        }

        fun invalidate() {
            field = Field.IGNORED
            command.clear()
            target = ByteArray(0)
            targetSize = 0
            targetOversized = false
        }

        fun reset() {
            field = Field.COMMAND
            command.clear()
            target = ByteArray(0)
            targetSize = 0
            targetOversized = false
        }

        private fun appendTarget(value: Byte) {
            if (targetOversized) return
            if (targetSize >= MAXIMUM_TARGET_BYTES) {
                target = ByteArray(0)
                targetSize = 0
                targetOversized = true
                return
            }
            if (targetSize == target.size) target = target.copyOf((target.size * 2).coerceAtLeast(64))
            target[targetSize++] = value
        }
    }

    private class PlainAttachLinkScanner {
        private val recent = ArrayList<Byte>(LONGEST_SCHEME_BYTES)
        private var candidate = ByteArray(0)
        private var candidateSize = 0
        private var oversized = false

        fun receive(value: Int): String? {
            if (isBoundary(value)) {
                val found = finishCandidate()
                recent.clear()
                return found
            }
            if (candidateSize > 0 || oversized) {
                appendCandidate(value.toByte())
                return null
            }
            recent += value.toByte()
            while (recent.size > LONGEST_SCHEME_BYTES) recent.removeAt(0)
            val scheme = WEB_SCHEMES.firstOrNull { recentEndsWith(it) } ?: return null
            candidate = scheme.copyOf()
            candidateSize = scheme.size
            recent.clear()
            return null
        }

        fun finishOutput(): String? {
            recent.clear()
            return finishCandidate()
        }

        fun reset() {
            recent.clear()
            candidate = ByteArray(0)
            candidateSize = 0
            oversized = false
        }

        private fun appendCandidate(value: Byte) {
            if (oversized) return
            if (candidateSize >= MAXIMUM_TARGET_BYTES) {
                candidate = ByteArray(0)
                candidateSize = 0
                oversized = true
                return
            }
            if (candidateSize == candidate.size) candidate = candidate.copyOf((candidate.size * 2).coerceAtLeast(64))
            candidate[candidateSize++] = value
        }

        private fun finishCandidate(): String? {
            if (oversized) {
                candidate = ByteArray(0)
                candidateSize = 0
                oversized = false
                return null
            }
            val result = runCatching { candidate.copyOf(candidateSize).decodeToString() }.getOrNull()
            candidate = ByteArray(0)
            candidateSize = 0
            return result?.let(::removeSurroundingPunctuation)?.takeIf(String::isNotEmpty)
        }

        private fun recentEndsWith(suffix: ByteArray): Boolean {
            if (recent.size < suffix.size) return false
            val offset = recent.size - suffix.size
            return suffix.indices.all {
                lowerAscii(recent[offset + it].toInt() and 0xFF) == (suffix[it].toInt() and 0xFF)
            }
        }

        private fun removeSurroundingPunctuation(text: String): String {
            var result = text
            while (result.isNotEmpty()) {
                when (result.last()) {
                    '.', ',', ';', ':', '!' -> result = result.dropLast(1)
                    ')' -> if (result.count { it == ')' } > result.count { it == '(' }) result = result.dropLast(1) else return result
                    ']' -> if (result.count { it == ']' } > result.count { it == '[' }) result = result.dropLast(1) else return result
                    '}' -> if (result.count { it == '}' } > result.count { it == '{' }) result = result.dropLast(1) else return result
                    else -> return result
                }
            }
            return result
        }

        private fun isBoundary(value: Int): Boolean =
            value <= 0x20 || value == 0x7F || value == '"'.code || value == '\''.code || value == '<'.code || value == '>'.code

        private fun lowerAscii(value: Int): Int = if (value in 'A'.code..'Z'.code) value + 0x20 else value

        private companion object {
            const val LONGEST_SCHEME_BYTES = 8
            val WEB_SCHEMES = arrayOf("http://".encodeToByteArray(), "https://".encodeToByteArray())
        }
    }

    private companion object {
        const val MAXIMUM_LINK_COUNT = 20
        const val MAXIMUM_TARGET_BYTES = 32 * 1024
        const val MAXIMUM_COMMAND_BYTES = 16
    }
}
